src/cim/components/ClusterDeployment/ClusterDeploymentHostSelectionStep.tsx

```diff
@@ -42,78 +50,90 @@ const getInitialValues = ({
-    agentLabels: agentSelector?.labels || [], // labelsToArray(agentSelector?.matchLabels),
-    locations: agentSelector?.locations || [], // getLocationsFromMatchExpressions(agentSelector?.matchExpressions),
-    selectedHostIds,
-
-    // Read-only, hidden for the user, not meant to be modified by the form
-    isSNOCluster,
+    agentLabels: agentSelector?.labels || [],
+    locations: agentSelector?.locations || [],
+    selectedHostIds: autoSelectHosts ? [] : selectedIds,
+    autoSelectedHostIds: autoSelectHosts ? selectedIds : [],
   };
 };
 
-const getValidationSchema = () =>
-  Yup.object().shape({
-    hostCount: hostCountValidationSchema,
-    useMastersAsWorkers: Yup.boolean().required(),
-    agentLabels: hostLabelsValidationSchema,
-    locations: Yup.array().min(1),
+const getValidationSchema = (agentClusterInstall: AgentClusterInstallK8sResource) => {
+  const isSNOCluster = agentClusterInstall?.spec?.provisionRequirements?.controlPlaneAgents === 1;
+
+  return Yup.lazy<ClusterDeploymentHostsSelectionValues>((values) => {
+    return Yup.object<ClusterDeploymentHostsSelectionValues>().shape({
+      hostCount: hostCountValidationSchema,
+      useMastersAsWorkers: Yup.boolean().required(),
+      autoSelectedHostIds: values.autoSelectHosts
+        ? Yup.array<string>().min(values.hostCount)
+        : Yup.array<string>(),
+      selectedHostIds: values.autoSelectHosts
+        ? Yup.array<string>()
+        : Yup.array<string>().min(isSNOCluster ? 1 : values.selectedHostIds.length === 4 ? 5 : 3),
+    });
   });
+};
 
 const ClusterDeploymentHostSelectionStep: React.FC<ClusterDeploymentHostSelectionStepProps> = ({
   clusterDeployment,
   agentClusterInstall,
-  // agents,
-  selectedHostIds,
-  matchingAgents: unfilteredMatchingAgents,
+  agents,
   onClose,
   onSaveHostsSelection,
-  ...rest
+  hostActions,
 }) => {
   const { addAlert } = useAlerts();
   const { setCurrentStepId } = React.useContext(ClusterDeploymentWizardContext);
 
   const initialValues = React.useMemo(
-    () => getInitialValues({ clusterDeployment, agentClusterInstall, selectedHostIds }),
+    () => getInitialValues({ agents, clusterDeployment, agentClusterInstall }),
     [], // eslint-disable-line react-hooks/exhaustive-deps
   );
-  const validationSchema = React.useMemo(() => getValidationSchema(), []);
 
-  // Use Ready hosts only
-  const matchingAgents = React.useMemo(
+  const validationSchema = React.useMemo(() => getValidationSchema(agentClusterInstall), [
+    agentClusterInstall,
+  ]);
+
+  const cdName = clusterDeployment?.metadata?.name;
+  const cdNamespace = clusterDeployment?.metadata?.namespace;
+
+  const availableAgents = React.useMemo(
     () =>
-      (unfilteredMatchingAgents || []).filter((agent) => {
+      (agents || []).filter((agent) => {
         const [status] = getAgentStatus(agent);
```

Comment:
We should filter-out agents based on `clusterDeploymentName` as well. The labels are set within UI flow only.

Reply:
https://github.com/rawagner/dynamic-cim/pull/51/files#diff-234124f695ab67686e1b6bdfdc04f939f8f90dcd18a76d6af459e516b28c0838L256

Reply:
why cannot we reassign agents from one cluster to another if they are not reserved ?